### PR TITLE
Universal checkpoint for zero stage 3

### PR DESCRIPTION
**Pull request by xylian86**
This PR enables the universal checkpoint for zero stage 3. 

Notes:

- The current implementation supports Data parallelism.
- Development is ongoing for universal checkpoint Stage 3 with tensor-slicing model parallelism.
- Pipeline parallelism is not supported by ZeRO Stage 3, and hence is not included in this universal checkpoint implementation.

In this PR:

- I've updated `deepspeed/checkpoint/ds_to_universal.py ` to support converting Zero checkpoints into Universal checkpoints.
- I've updated `deepspeed/runtime/zero/stage3.py` to enable loading Universal checkpoints using the Stage 3 optimizer.

### CODE REVIEW

**Comment by xylian86** [comment]
@microsoft-github-policy-service agree

On Apr 29, 2024, at 12:59 AM, microsoft-github-policy-service[bot] ***@***.***> wrote:



@xylian86<https://urldefense.com/v3/__https://github.com/xylian86__;!!DZ3fjg!_mO4Jru0mfXYSDGJdXwCfSsKmEA27OGalsA2H3x7ihu18b2EIw_RwoT3HcxfEPsBtVoDtyGn60tvuApoA9AD37ScuA$> please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.

@microsoft-github-policy-service agree [company="{your company}"]


Options:

  *   (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.

@microsoft-github-policy-service agree


  *   (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.

@microsoft-github-policy-service agree company="Microsoft"


Contributor License Agreement
Contribution License Agreement

This Contribution License Agreement (“Agreement”) is agreed to by the party signing below (“You”),
and conveys certain license rights to Microsoft Corporation and its affiliates (“Microsoft”) for Your
contributions to Microsoft open source projects. This Agreement is effective as of the latest signature
date below.

  1.  Definitions.
“Code” means the computer software code, whether in human-readable or machine-executable form,
that is delivered by You to Microsoft under this Agreement.
“Project” means any of the projects owned or managed by Microsoft and offered under a license
approved by the Open Source Initiative (www.opensource.org<https://urldefense.com/v3/__http://www.opensource.org__;!!DZ3fjg!_mO4Jru0mfXYSDGJdXwCfSsKmEA27OGalsA2H3x7ihu18b2EIw_RwoT3HcxfEPsBtVoDtyGn60tvuApoA9B65UDdDw$>).
“Submit” is the act of uploading, submitting, transmitting, or distributing code or other content to any
Project, including but not limited to communication on electronic mailing lists, source code control
systems, and issue tracking systems that are managed by, or on behalf of, the Project for the purpose of
discussing and improving that Project, but excluding communication that is conspicuously marked or
otherwise designated in writing by You as “Not a Submission.”
“Submission” means the Code and any other copyrightable material Submitted by You, including any
associated comments and documentation.
  2.  Your Submission. You must agree to the terms of this Agreement before making a Submission to any
Project. This Agreement covers any and all Submissions that You, now or in the future (except as
described in Section 4 below), Submit to any Project.
  3.  Originality of Work. You represent that each of Your Submissions is entirely Your original work.
Should You wish to Submit materials that are not Your original work, You may Submit them separately
to the Project if You (a) retain all copyright and license information that was in the materials as You
received them, (b) in the description accompanying Your Submission, include the phrase “Submission
containing materials of a third party:” followed by the names of the third party and any licenses or other
restrictions of which You are aware, and (c) follow any other instructions in the Project’s written
guidelines concerning Submissions.
  4.  Your Employer. References to “employer” in this Agreement include Your employer or anyone else
for whom You are acting in making Your Submission, e.g. as a contractor, vendor, or agent. If Your
Submission is made in the course of Your work for an employer or Your employer has intellectual
property rights in Your Submission by contract or applicable law, You must secure permission from Your
employer to make the Submission before signing this Agreement. In that case, the term “You” in this
Agreement will refer to You and the employer collectively. If You change employers in the future and
desire to Submit additional Submissions for the new employer, then You agree to sign a new Agreement
and secure permission from the new employer before Submitting those Submissions.
  5.  Licenses.

  *   Copyright License. You grant Microsoft, and those who receive the Submission directly or
indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license in the
Submission to reproduce, prepare derivative works of, publicly display, publicly perform, and distribute
the Submission and such derivative works, and to sublicense any or all of the foregoing rights to third
parties.
  *   Patent License. You grant Microsoft, and those who receive the Submission directly or
indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license under
Your patent claims that are necessarily infringed by the Submission or the combination of the
Submission with the Project to which it was Submitted to make, have made, use, offer to sell, sell and
import or otherwise dispose of the Submission alone or with the Project.
  *   Other Rights Reserved. Each party reserves all rights not expressly granted in this Agreement.
No additional licenses or rights whatsoever (including, without limitation, any implied licenses) are
granted by implication, exhaustion, estoppel or otherwise.

  1.  Representations and Warranties. You represent that You are legally entitled to grant the above
licenses. You represent that each of Your Submissions is entirely Your original work (except as You may
have disclosed under Section 3). You represent that You have secured permission from Your employer to
make the Submission in cases where Your Submission is made in the course of Your work for Your
employer or Your employer has intellectual property rights in Your Submission by contract or applicable
law. If You are signing this Agreement on behalf of Your employer, You represent and warrant that You
have the necessary authority to bind the listed employer to the obligations contained in this Agreement.
You are not expected to provide support for Your Submission, unless You choose to do so. UNLESS
REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING, AND EXCEPT FOR THE WARRANTIES
EXPRESSLY STATED IN SECTIONS 3, 4, AND 6, THE SUBMISSION PROVIDED UNDER THIS AGREEMENT IS
PROVIDED WITHOUT WARRANTY OF ANY KIND, INCLUDING, BUT NOT LIMITED TO, ANY WARRANTY OF
NONINFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
  2.  Notice to Microsoft. You agree to notify Microsoft in writing of any facts or circumstances of which
You later become aware that would make Your representations in this Agreement inaccurate in any
respect.
  3.  Information about Submissions. You agree that contributions to Projects and information about
contributions may be maintained indefinitely and disclosed publicly, including Your name and other
information that You submit with Your Submission.
  4.  Governing Law/Jurisdiction. This Agreement is governed by the laws of the State of Washington, and
the parties consent to exclusive jurisdiction and venue in the federal courts sitting in King County,
Washington, unless no federal subject matter jurisdiction exists, in which case the parties consent to
exclusive jurisdiction and venue in the Superior Court of King County, Washington. The parties waive all
defenses of lack of personal jurisdiction and forum non-conveniens.
  5.  Entire Agreement/Assignment. This Agreement is the entire agreement between the parties, and
supersedes any and all prior agreements, understandings or communications, written or oral, between
the parties relating to the subject matter hereof. This Agreement may be assigned by Microsoft.

File: deepspeed/runtime/zero/stage3.py [hunk 1]
```diff
@@ -2622,24 +2623,105 @@ def load_state_dict(self,
...
+                    dist.send(tensor=load_serial, dst=rank + 1)
+
+            if len(self.persistent_parameters) > 0:
+                self.persistent_parameters[0].partition(self.persistent_parameters)
+                # self.persistent_parameters[0].all_gather(self.persistent_parameters) # this will be done in checkpoint_event_epilogue() so remove it to prevent double all_gather
+
+    def _load_universal_checkpoint(self, checkpoint_folder, load_optimizer_states, load_from_fp32_weights, param_shapes):
+        self.load_checkpoint_state_from_checkpoint_dir_stage3(checkpoint_folder, param_shapes)
+
+    def load_hp_checkpoint_state_from_checkpoint_dir_stage3(self, checkpoint_dir, param_shapes):
```
Comment (tohtana) [inline deepspeed/runtime/zero/stage3.py:2651]: We have similar function for ZeRO 1/2. Can we extract common parts to avoid duplication?

File: deepspeed/checkpoint/ds_to_universal.py [hunk 2]
```diff
@@ -354,38 +466,63 @@ def main(args):
...
-        mp_sd = torch.load(mp_rank_file, map_location=torch.device('cpu'))
-        slice_shapes += mp_sd[PARAM_SHAPES]
+        # fix back to normal flat dict, merge duplicates for tp>1
+        slice_shapes = dict((k, v) for d in slice_shapes for k, v in d.items())
+        temp_dir = os.path.join(args.output_folder, 'tmp')
 
-    # fix back to normal flat dict, merge duplicates for tp>1
-    slice_shapes = dict((k, v) for d in slice_shapes for k, v in d.items())
-    temp_dir = os.path.join(args.output_folder, 'tmp')
+        print('*** 1. Extracting ZeRO fragments')
```
Comment (tohtana) [inline deepspeed/checkpoint/ds_to_universal.py:490]: Can we show the progress as we do for ZeRO 1/2?

File: deepspeed/checkpoint/ds_to_universal.py [hunk 3]
```diff
@@ -70,6 +75,19 @@ def parse_arguments():
     return args
 
 
+def atoi(text):
+    return int(text) if text.isdigit() else text
+
+
+def natural_keys(text):
```
Comment (tohtana) [inline deepspeed/checkpoint/ds_to_universal.py:82]: Thank you for introducing this interesting approach.
We have a similar sorting in `_merge_zero_shards` but are using a different approach but it is not good to have two different sorting implementations for the same purpose. Can you replace this one with `natural_keys`?

**Review by tohtana — COMMENTED** [review]
We have a [test](https://github.com/microsoft/DeepSpeed/blob/0e6fcdb84b3690ff45cb0fc6b1945f0d97f0d830/tests/unit/checkpoint/test_universal_checkpoint.py#L165) for universal checkpointing. It currently supports DP scaling only but it would be good to test ZeRO3 feature using this test. You can just add "3" to the test argument.

File: deepspeed/checkpoint/ds_to_universal.py [hunk 4]
```diff
@@ -332,6 +350,70 @@ def _merge_tp_slice_files(args, ds_checkpoint, slice_shapes, temp_dir):
...
+    total_numel = 0
+    total_params = 0
+
+    for name, shape in param_shapes.items():
+        unpartitioned_numel = shape.numel()
+        total_numel += unpartitioned_numel
+        total_params += 1
+
+        partitioned_numel, _ = _zero_partitioned_param_info(unpartitioned_numel, world_size)
+        print(key_tensors[0])
```
[the remark this answers is not in the document]
Comment (xylian86) [reply deepspeed/checkpoint/ds_to_universal.py]: Thanks for the suggestion and sorry for keeping the debug statement. I have removed this line in the new commit.

File: deepspeed/runtime/zero/stage3.py [hunk 5]
```diff
@@ -2622,24 +2623,105 @@ def load_state_dict(self,
...
+                rank_end = dist.get_world_size() - 1
+                if local_rank != rank_end:
+                    dist.send(tensor=load_serial, dst=rank + 1)
+
+            if len(self.persistent_parameters) > 0:
+                self.persistent_parameters[0].partition(self.persistent_parameters)
+                # self.persistent_parameters[0].all_gather(self.persistent_parameters) # this will be done in checkpoint_event_epilogue() so remove it to prevent double all_gather
+
+    def _load_universal_checkpoint(self, checkpoint_folder, load_optimizer_states, load_from_fp32_weights, param_shapes):
+        self.load_checkpoint_state_from_checkpoint_dir_stage3(checkpoint_folder, param_shapes)
```
[the remark this answers is not in the document]
Comment (xylian86) [reply deepspeed/runtime/zero/stage3.py]: Yes, I figured out that the function name was not updated when the code was merged. I have fixed it in the new commit.

File: deepspeed/runtime/zero/stage3.py (hunk 1 above)
Comment (xylian86) [reply deepspeed/runtime/zero/stage3.py:2651]: Hi @tohtana, thanks for the suggestion!
For those two `similar` functions for Zero 1/2 and Zero 3, due to the different sharing mechanisms (one is shared across layers while the other is shared within each layer), the implementation of these two functions varies considerably, making it challenging to extract common elements. Please let me know if you have additional suggestions.

File: deepspeed/checkpoint/ds_to_universal.py (hunk 3 above)
Comment (xylian86) [reply deepspeed/checkpoint/ds_to_universal.py:82]: Thanks for the suggestions!
For this `natural_keys` function, I actually reuse it from [zero_to_fp32.py](https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/utils/zero_to_fp32.py#L52).

You’re right; it’s not ideal to have two different implementations for the same function. How about I replace the one in `_merge_zero_shards` with this `natural_keys`?

File: deepspeed/checkpoint/ds_to_universal.py (hunk 2 above)
Comment (xylian86) [reply deepspeed/checkpoint/ds_to_universal.py:489]: For sure. I add the print info in the new commit. Note that the current ZeRO 3 only supports DP, therefore, the first step `Extracting ZeRO fragments` is skipped, unlike in ZeRO 1/2.

**Comment by xylian86** [comment]
image

Convergence curve for ZeRO 3 using the current implementation

File: deepspeed/checkpoint/ds_to_universal.py (hunk 3 above)
Comment (tohtana) [reply deepspeed/checkpoint/ds_to_universal.py:79]: This sounds good, thank you

**Review by tohtana — APPROVED** [review]

**Comment by Lomax314** [comment]
Hello, thank you for your work!
I've encountered some issues during my practice and would appreciate your help in analyzing them.
I am using Zero Stage=3 for training and use model.save_checkpoint to save the optimizer state(optimizer = torch.optim.AdamW(model.parameters(), lr=args.lr)). The conversion of the original optimizer state to the universal_checkpoint format and loading the uni format part did not report any errors. However, the training loss did not continue from where it left off but started to gradually decrease from 9, which is a very high loss value. I am puzzled by this and would appreciate your assistance in analyzing it.

Additionally, I used model.state_dict() and uni_model.state_dict() to get the model state before and after the conversion and compared them using compare_state_dicts. The result triggered an assertion, as torch.equal(s0.to('cpu'), s1.to('cpu')) evaluated to False.

**Comment by ArtificialZeng** [comment]
> This PR enables the universal checkpoint for zero stage 3.
> 
> Notes:
> 
> * The current implementation supports Data parallelism.
> * Development is ongoing for universal checkpoint Stage 3 with tensor-slicing model parallelism.
> * Pipeline parallelism is not supported by ZeRO Stage 3, and hence is not included in this universal checkpoint implementation.
> 
> In this PR:
> 
> * I've updated `deepspeed/checkpoint/ds_to_universal.py ` to support converting Zero checkpoints into Universal checkpoints.
> * I've updated `deepspeed/runtime/zero/stage3.py` to enable loading Universal checkpoints using the Stage 3 optimizer.

how to load

**Comment by xylian86** [comment]
@ArtificialZeng We have released examples in the Megatron-DeepSpeed repository. You can find them at:
https://github.com/microsoft/Megatron-DeepSpeed/tree/main/examples_deepspeed/universal_checkpointing.
Please let us know if you encounter any issues.
